src/main/org/apache/tools/ant/taskdefs/optional/junitlauncher/JUnitLauncherTask.java

```diff
@@ -0,0 +1,508 @@
+                    final List<TestExecutionListener> testExecutionListeners = new ArrayList<>();
+                    // a listener that we always put at the front of list of listeners
+                    // for this request.
+                    final Listener firstListener = new Listener();
+                    // we always enroll the summary generating listener, to the request, so that we
+                    // get to use some of the details of the summary for our further decision making
+                    testExecutionListeners.add(firstListener);
+                    testExecutionListeners.addAll(getListeners(testRequest, executionCL));
+                    final PrintStream originalSysOut = System.out;
+                    final PrintStream originalSysErr = System.err;
+                    try {
+                        firstListener.switchedSysOutHandle = trySwitchSysOut(testRequest);
+                        firstListener.switchedSysErrHandle = trySwitchSysErr(testRequest);
+                        launcher.execute(request, testExecutionListeners.toArray(new TestExecutionListener[testExecutionListeners.size()]));
+                    } finally {
+                        // switch back sysout/syserr to the original
+                        try {
+                            System.setOut(originalSysOut);
+                        } catch (Exception e) {
+                            // ignore
+                        }
+                        try {
+                            System.setErr(originalSysErr);
+                        } catch (Exception e) {
+                            // ignore
+                        }
+                    }
+                    handleTestExecutionCompletion(test, firstListener.getSummary());
+                } finally {
+                    try {
+                        testRequest.close();
+                    } catch (Exception e) {
+                        // log and move on
+                        log("Failed to cleanly close test request", e, Project.MSG_DEBUG);
+                    }
+                }
+            }
+        } finally {
+            Thread.currentThread().setContextClassLoader(previousClassLoader);
+        }
+    }
+
+    /**
+     * @return Creates and returns the a {@link Path} which will be used as the classpath of this
+     * task. This classpath will then be used for execution of the tests
+     */
+    public Path createClassPath() {
+        this.classPath = new Path(getProject());
+        return this.classPath;
+    }
+
+    /**
+     * @return Creates and returns a {@link SingleTestClass}. This test will be considered part of the
+     * tests that will be passed on to the underlying JUnit platform for possible execution of the test
+     */
+    public SingleTestClass createTest() {
+        final SingleTestClass test = new SingleTestClass();
+        this.preConfigure(test);
```

Comment:
`create*` is called before the attribute setters, If you call `preconfigure` here the test won't see the actual values set at the task level.

`addConfigured` would get called after the attribute setters, alternatively you need to defer pre-configuration until `execute` is called. With both approaches you need to ensure you don't overwrite the test's explicit configuration with values from the task, of course.

Same for `createTestClasses`.

Reply:
Thank you, for catching this. I went back and read the Ant manual about writing custom tasks (which I had skipped the first time around) and now understand this logic of execution better. I have updated the PR to fix this and use the `addConfiguredXXX` construct and also made sure my `preConfigure` implementation doesn't overwrite values that might have been specified on the nested elements.